src/securityPolicy.c

```diff
@@ -565,53 +640,99 @@ security_policy_t policyForSignTxStakePoolRegistrationConfirm(
 }
 
 // For each withdrawal
-security_policy_t policyForSignTxWithdrawal()
+security_policy_t policyForSignTxWithdrawal(
+        sign_tx_signingmode_t txSigningMode,
+        const stake_credential_t* stakeCredential
+)
 {
-	// No need to check withdrawals
-	SHOW();
-}
+	switch (txSigningMode) {
+	case SIGN_TX_SIGNINGMODE_ORDINARY_TX:
+		DENY_UNLESS(stakeCredential->type == STAKE_CREDENTIAL_KEY_PATH);
+		DENY_UNLESS(bip44_isOrdinaryStakingKeyPath(&stakeCredential->keyPath));
+		ALLOW();
+		break;
 
+	case SIGN_TX_SIGNINGMODE_SCRIPT_TX:
+		DENY_UNLESS(stakeCredential->type == STAKE_CREDENTIAL_SCRIPT_HASH);
+		ALLOW();
+		break;
 
-// For each transaction witness
-// Note: witnesses reveal public key of an address
-// and Ledger *does not* check whether they correspond to
-// previously declared inputs and certificates
-security_policy_t policyForSignTxWitness(
-        sign_tx_usecase_t signTxUsecase,
-        const bip44_path_t* pathSpec
-)
-{
-	switch (signTxUsecase) {
+	default:
+		ASSERT(false);
+	}
 
-	case SIGN_TX_USECASE_ORDINARY_TX: {
+	DENY(); // should not be reached
+}
 
-		switch (bip44_classifyPath(pathSpec)) {
+static inline security_policy_t ordinaryWitnessPolicy(const bip44_path_t* pathSpec, bool mintPresent)
+{
+	switch (bip44_classifyPath(pathSpec)) {
+	case PATH_ORDINARY_SPENDING_KEY:
+	case PATH_ORDINARY_STAKING_KEY:
+	case PATH_POOL_COLD_KEY:
+		if (bip44_isPathReasonable(pathSpec)) {
+			ALLOW();
+		} else {
+			WARN();
+		}
+		break;
 
-		case PATH_WALLET_SPENDING_KEY:
-		case PATH_WALLET_STAKING_KEY:
-		case PATH_POOL_COLD_KEY:
-			if (bip44_isPathReasonable(pathSpec)) {
-				ALLOW();
-			} else {
-				WARN();
-			}
-			break;
+	case PATH_MINT_KEY:
+		DENY_UNLESS(mintPresent);
+		SHOW();
+		break;
 
-		default:
-			DENY();
-			break;
+	default:
+		DENY();
+		break;
+	}
+}
+
+static inline security_policy_t multisigWitnessPolicy(const bip44_path_t* pathSpec, bool mintPresent)
+{
+	switch (bip44_classifyPath(pathSpec)) {
+	case PATH_MULTISIG_SPENDING_KEY:
+	case PATH_MULTISIG_STAKING_KEY:
+	case PATH_POOL_COLD_KEY:
+		if (bip44_isPathReasonable(pathSpec)) {
+			SHOW();
+		} else {
+			WARN();
 		}
+		break;
+
+	case PATH_MINT_KEY:
+		DENY_UNLESS(mintPresent);
+		SHOW();
 
+	default:
+		DENY();
 		break;
 	}
+}
+
+// For each transaction witness
+// Note: witnesses reveal public key of an address
+// and Ledger *does not* check whether they correspond to
+// previously declared inputs and certificates
+security_policy_t policyForSignTxWitness(
+        sign_tx_signingmode_t txSigningMode,
+        const bip44_path_t* pathSpec,
+        bool mintPresent
+)
+{
+	switch (txSigningMode) {
+	case SIGN_TX_SIGNINGMODE_ORDINARY_TX:
+		return ordinaryWitnessPolicy(pathSpec, mintPresent);
 
-	case SIGN_TX_USECASE_POOL_REGISTRATION_OWNER: {
+	case SIGN_TX_SIGNINGMODE_SCRIPT_TX:
+		return multisigWitnessPolicy(pathSpec, mintPresent);
 
+	case SIGN_TX_SIGNINGMODE_POOL_REGISTRATION_OWNER: {
```

Comment:
Perhaps the two remaining case statements could also be extracted out into functions?